### ISSUE TITLE
Integrate deep link into `url` command

### ISSUE THREAD

**Issue by christian-bromann** [webdriverio/webdriverio#6538]
Using deeplinks in mobile automation can be a very useful feature as it allows to open specific views of an app with one single call rather than walking through the app over and over again. I suggest to enhance to enhance the `url` command to support this functionality from the [`wdio-mobile-utils`](https://github.com/martinfrancois/wdio-mobile-utils/blob/master/src/lib/deeplink.ts) package with the following interface:

- before: `url(path: string)`
- after: `url(path: string, urlOpts: UrlCommandOptions)`
  - where `UrlCommandOptions = { id: string }`
  - where `id` is the bundle or app id depending which environment the user is using
  - the original implementation comes with a timeout option for its iOS implementation, I suggest to make a good guess here and hard code this timeout to avoid to many options that only confuse users and aren't gonna be used much anyway

**Comment by wswebcreation** [comment]
I need to do some investigation on this, but what if WebdriverIO could determine the bundleID/package name with an Appium command and check the url based on a regular expression? This might make it easier to use and prevents extra arguments. What do you think @christian-bromann ?

**Comment by christian-bromann** [comment]
Ah so you are suggesting if someone would call:

```js
browser.url('swaglabs://swag-overview/0-1')
```

We are able to determine the app/bundle id from the url `swaglabs://`? That would be great!

**Comment by wswebcreation** [comment]
We can't determine it from the url, but afaik we get a lot of information back in the driver. As said, I need to investigate if the bundleID/packageName is already in there (or in a different place). This means that the url command needs to determine this with internal data from the driver, or with an extra Appium command.

**Comment by jayandran-Sampath** [comment]
Hi, any help required on this issue? 

Did the approach is finalised?
@wswebcreation  @christian-bromann

**Comment by christian-bromann** [comment]
@jayandran-Sampath any help on this issue would be very appreciated.

**Comment by jayandran-Sampath** [comment]
Hey @christian-bromann  : Sorry its been a while to work on this task. I did basic analysis on how appium implemented deeplinking.. My finding is that we dont need to do anything specially in Webdriverio. Appium internally takes cares of Deeplinking in case of its native app. Just we need to use /url api with parameter as deeplink url. 

I tried with simple android app with webdriverio.. it works fine 

Appium Logs : 

[35m[HTTP][39m [37m-->[39m [37mPOST[39m [37m/session/dd87b9e1-dd53-4955-bd8f-37ebaa950073/url[39m
[35m[HTTP][39m [90m{"url":"theapp://login/alice/mypassword"}[39m
[debug] [35m[W3C (dd87b9e1)][39m Calling AppiumDriver.setUrl() with args: ["theapp://login/alice/mypassword","dd87b9e1-dd53-4955-bd8f-37ebaa950073"]
[debug] [35m[ADB][39m Running ‘~/Android/sdk/platform-tools/adb -P 5037 -s emulator-5554 shell am start -W -a android.intent.action.VIEW -d theapp://login/alice/mypassword io.cloudgrey.the_app'

Wdio Logs : 

2021-09-17T09:01:28.745Z INFO webdriver: COMMAND navigateTo("theapp://login/alice/mypassword")
2021-09-17T09:01:28.745Z INFO webdriver: [POST] http://localhost:4723/session/dd87b9e1-dd53-4955-bd8f-37ebaa950073/url
2021-09-17T09:01:28.745Z INFO webdriver: DATA { url: 'theapp://login/alice/mypassword' }

**Comment by wswebcreation** [comment]
Not 100% correct. This will work on Android Emulators and real devices, on iOS sims, but not on iOS real devices. That command will open Siri.
There is a workaround which is used in the Appium boilerplate

**Comment by jayandran-Sampath** [comment]
yes @wswebcreation  i have also gone through the appium pro blog..  but my understanding is that this issue is created only to handle the deeplinking with same command browser.url rather than using other way (https://github.com/martinfrancois/wdio-mobile-utils/blob/master/src/lib/deeplink.ts)

And that is implicitly taken care by appium so no need to work on it in webdriverio

**Comment by wswebcreation** [comment]
My suggestion would be to create a new command to prevent complexity and call the `browser.url('string')` internally for all devices instead of iOS real devices (as long as we can determine based on the capabilities that we know it's a real iOS device)

**Comment by jayandran-Sampath** [comment]
i could not find any straight forward capability to determine whether its simulator or real device. One way i was thinking to verify based on UUID Pattern. I analysed few ios real devices where all UUID are starting with 8 characters followed by hypen and characters where as Simulators have more than one hypen like below. if we have can put some matcher based on this condition.. we can identify whether its real device or simulator. Also im not sure whether its 100% accurate as well

Real Devices : XXXXXXXX-XXXXXXXXXX

Simulators : XXXXXXXX-XXXX-XXXX-XXXXX

Please share your thoughts too

**Comment by christian-bromann** [comment]
> My suggestion would be to create a new command

I would be against creating a new command. WebdriverIO already has a lot and there is not much value in it. In the `url` command we should detect for iOS real devices and apply a different behavior there (in case a deeplink url is given).

**Comment by wswebcreation** [comment]
> i could not find any straight forward capability to determine whether its simulator or real device. One way i was thinking to verify based on UUID Pattern. I analysed few ios real devices where all UUID are starting with 8 characters followed by hypen and characters where as Simulators have more than one hypen like below. if we have can put some matcher based on this condition.. we can identify whether its real device or simulator. Also im not sure whether its 100% accurate as well
> 
> Real Devices : XXXXXXXX-XXXXXXXXXX
> 
> Simulators : XXXXXXXX-XXXX-XXXX-XXXXX
> 
> Please share your thoughts too

Hmm, that's a good one @jayandran-Sampath , I'll also explore it and will test this for the appium-boilerplate

**Comment by wswebcreation** [comment]
Ok, this might be useful 

- https://blog.diawi.com/2018/10/15/2018-apple-devices-and-their-new-udid-format/
- https://www.theiphonewiki.com/wiki/UDID

Based on that we can determine that if it has 2 or more `-` something like this

```js
const regex = new RegExp('(.*-.*){2,}');
// emulator
console.log(regex.test('682A72C1-C8AB-4AF9-97B5-3D03CA31A7CC'));
// expected output: true

// Real device after 2018
console.log(regex.test('00008020-0012345A1234567B'));
// expected output: false

// real device before 2018
console.log(regex.test('123456789abcdef0123456789abcdef012345678'));
// expected output: true
```

What do you think?

**Comment by christian-bromann** [comment]
Make sense to me!

**Comment by wswebcreation** [comment]
Ok, then I'll start playing with in the appium-boilerplate and we can then copy the code over if it keeps working stabile

**Comment by wswebcreation** [comment]
ok, tested it with the appium-boilerplate with the old implementation on sims with opening Safari, entering the address and get to the deeplink screen

```log
[iPhone 12 iOS 14.5 #0-0] Running: iPhone 12 on iOS 14.5 executing /Users/wimselles/Sauce/Git/appium-boilerplate/apps/iOS-Simulator-NativeDemoApp-0.4.0.app.zip
[iPhone 12 iOS 14.5 #0-0] Session ID: 9dca563b-3124-480f-8779-93c3f1fd4533
[iPhone 12 iOS 14.5 #0-0]
[iPhone 12 iOS 14.5 #0-0] » /tests/specs/app.deep.link.navigation.spec.ts
[iPhone 12 iOS 14.5 #0-0] WebdriverIO and Appium, when navigating by deep link
[iPhone 12 iOS 14.5 #0-0]    ✓ should be able to open the webview
[iPhone 12 iOS 14.5 #0-0]    ✓ should be able to open the login form screen
[iPhone 12 iOS 14.5 #0-0]    ✓ should be able to open the forms screen
[iPhone 12 iOS 14.5 #0-0]    ✓ should be able to open the swipe screen
[iPhone 12 iOS 14.5 #0-0]    ✓ should be able to open the drag and drop screen
[iPhone 12 iOS 14.5 #0-0]    ✓ should be able to open the home screen
[iPhone 12 iOS 14.5 #0-0]
[iPhone 12 iOS 14.5 #0-0] 6 passing (56.6s)
```

Same now with the new way, meaning determine that you are a sim and open it direct from the app with the `browser.url('deeplink://url')` and this is the result

```log
[iPhone 12 iOS 14.5 #0-0] Running: iPhone 12 on iOS 14.5 executing /Users/wimselles/Sauce/Git/appium-boilerplate/apps/iOS-Simulator-NativeDemoApp-0.4.0.app.zip
[iPhone 12 iOS 14.5 #0-0] Session ID: 9dca563b-3124-480f-8779-93c3f1fd4533
[iPhone 12 iOS 14.5 #0-0]
[iPhone 12 iOS 14.5 #0-0] » /tests/specs/app.deep.link.navigation.spec.ts
[iPhone 12 iOS 14.5 #0-0] WebdriverIO and Appium, when navigating by deep link
[iPhone 12 iOS 14.5 #0-0]    ✓ should be able to open the webview
[iPhone 12 iOS 14.5 #0-0]    ✓ should be able to open the login form screen
[iPhone 12 iOS 14.5 #0-0]    ✓ should be able to open the forms screen
[iPhone 12 iOS 14.5 #0-0]    ✓ should be able to open the swipe screen
[iPhone 12 iOS 14.5 #0-0]    ✓ should be able to open the drag and drop screen
[iPhone 12 iOS 14.5 #0-0]    ✓ should be able to open the home screen
[iPhone 12 iOS 14.5 #0-0]
[iPhone 12 iOS 14.5 #0-0] 6 passing (23.9s)
```

Almost half the time!!!!! 🎉

**Comment by jayandran-Sampath** [comment]
Nice @wswebcreation .

So we will modify wdio url function to determine whether session is for App (Based on bundle ID) & IOS Real Device (Based on udid pattern) and then we will return some message saying Deeplink Command is not supported by appium using get url. Use alternate steps

Right?

**Comment by christian-bromann** [comment]
I believe if users can use the `url` command to open a resource and it works for desktop and for mobile using deeplinks without them having to worry about how to implement this, it would be excellent.

**Comment by jayandran-Sampath** [comment]
@christian-bromann  Even i thought the same but had concerns on whether these locators would remain same for all versions? and also are we good to club multiple commands(execute,waitFordisplayed,click) within single function

**Comment by christian-bromann** [comment]
> are we good to club multiple commands(execute,waitFordisplayed,click) within single function

Yes. This is the idea behind commands within the WebdriverIO package. They should do abstract things within a single command.

**Comment by wswebcreation** [comment]
@christian-bromann 

I've been thinking about this one, and also about your remark

> I believe if users can use the url command to open a resource and it works for desktop and for mobile using deeplinks without them having to worry about how to implement this, it would be excellent.

but what if we add a `driver.deepLink(args)` command where all logic is only available for mobile. The deeplink command is currently not available as a method in Appium (only deeplink for Android as a `driver.execute('mobile:deeplink', args)` and it would keep the complexity at the level where it needs to be

**Comment by christian-bromann** [comment]
I guess `url` is a bit ambiguous here as we don't really open an url but a view. We should consider renaming it to `open`. Generally I would prefer to keep it (e.g. open a resource which can be an url a view or whatever) within a single command if possible. What do you think?

**Comment by wswebcreation** [comment]
I think I would then use the name `deepLink` instead of `open` because `deepLink` is more common in the mobile world. Otherwise we would introduce a new command which could cause confusion. `open` is also used in other frameworks to open a browser with a given url

**Comment by christian-bromann** [comment]
My concern here is that we have many commands that do somewhat a similar action. As WebdriverIO already has a very large set of commands I want to be cautious to not add more commands if no necessary. In this case, given the mobile context, adding a new `deepLink` command sounds good to me.

**Comment by wswebcreation** [comment]
Cool! Will put it on my agenda to work on it before the end of the year

**Comment by kopax-polyconseil** [comment]
Hello @wswebcreation, I am trying to open firebase dynamic links and universal links as a POC. 

I tried to open an URL on android using:

```bash
    return driver.execute('mobile:deepLink', {
      url: 'wdio://http://localhost:3000/signup-confirmation?token=xuZwRqTCsWyzMGcOgoaCLzS50o8VK_Y_qXVdpDFhWfg&expiration_timestamp=1672686846&email=dka%2B1672684902737%40passculture.app',
      package: `app.passculture.staging/com.passculture.MainActivity`,
    })
```

It open the app, but I stay on the home, not triggering navigation. I will of course debug, but if you have a clue, thank you !

I also hope this will be merged, as very useful. Thanks for sharing !

**Comment by wswebcreation** [comment]
Closed by https://github.com/webdriverio/webdriverio/pull/14212
